### PR TITLE
Make mobilecoind expose network fees via an API call

### PR DESCRIPTION
**Pull request by garbageslam**
This makes mobilecoind hit consensus' `get_last_block_info` as part
of the `GetNetworkStatus` call. It forwards a subset of that information
in response.

There is some argument that it should forward an entire `LastBlockInfo`
response object without modification -- I decided not to do that
because mobilecoind's proto doesn't currently depend on the consensus
api proto, only the `external.proto` and `blockchain.proto`. Feel
free to tell me to that dep is okay though.

This will allow mobilecoind python scripts and other clients of
mobilecoind to know the current (dynamic) network fees and block
version and such. Currently, there is no way for them to do that.

### CODE REVIEW

**Review by eranrund — COMMENTED** [review]
I think its fine to add the dependency and return the entire block info.
Once concern I have here is that before this change `get_network_status` would not do any RPC calls and as such would be very quick. This could potentially affect various scripts that poll this endpoint.
I am not sure there is an easy workaround but I think its worth exploring whether the block info can be plumbed from https://github.com/mobilecoinfoundation/mobilecoin/blob/master/ledger/sync/src/network_state/polling_network_state.rs#L92 (`fetch_block_height` uses the last block info call internally) since this is already polling the nodes.
It would be a bigger change, but result in a more efficient API.

**Comment by garbageslam** [comment]
@eranrund maybe we could just like, add another call to the mobilecoind api? do you think that would be alright?

**Comment by eranrund** [comment]
> @eranrund maybe we could just like, add another call to the mobilecoind api? do you think that would be alright?

that would also work.

**Comment by garbageslam** [comment]
Maybe it's not that difficult to do what you said, will look at it

**Review by samdealy — APPROVED** [review]

File: mobilecoind/src/payments.rs [hunk 1]
```diff
@@ -472,15 +466,16 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
     /// * `token_id` - The token id to transact in.
     /// * `inputs` - UTXOs that will be spent by the transaction.
     /// * `receiver` - The single receiver of the transaction's outputs.
-    /// * `fee` - Transaction fee. If zero, defaults to the highest fee set by
-    ///   configured consensus nodes, or the hard-coded FALLBACK_FEE.
+    /// * `opt_fee` - Transaction fee. If zero, defaults to the highest fee set
```
Comment (eranrund) [inline mobilecoind/src/payments.rs:469]: Need to add `last_block_infos`

File: mobilecoind/src/service.rs [hunk 2]
```diff
@@ -1927,6 +1950,26 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
         }
         let local_block_index = num_blocks - 1;
 
+        // Get LastBlockInfo from our peers
+        let block_infos = self.get_last_block_infos();
+        // choose the block info which is latest in terms of block index (we may be
+        // isolated from some of the nodes)
+        let last_block_info = if let Some(latest) =
+            block_infos.into_iter().reduce(|a, b| -> BlockInfo {
```
Comment (eranrund) [inline mobilecoind/src/service.rs:1958]: I think `block_infos.into_iter().map(|block_info| block_info.block_index).max()` is more readable

**Review by eranrund — APPROVED** [review]

File: connection/src/traits.rs [hunk 3]
```diff
@@ -116,6 +116,20 @@ impl From<LastBlockInfoResponse> for BlockInfo {
...
 
+impl From<BlockInfo> for LastBlockInfoResponse {
+    fn from(src: BlockInfo) -> Self {
+        let mut result = LastBlockInfoResponse::new();
+        result.index = src.block_index;
+        result.network_block_version = src.network_block_version;
+        result.set_minimum_fees(HashMap::from_iter(
+            src.minimum_fees
+                .iter()
+                .map(|(token_id, fee)| (**token_id, *fee)),
```
Comment (remoun) [inline connection/src/traits.rs:127]: Might as well use `into_iter` to consume the object

```suggestion
                .into_iter()
                .map(|(token_id, fee)| (*token_id, fee)),
```

File: mobilecoind/src/service.rs [hunk 4]
```diff
@@ -1927,6 +1950,20 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
...
+        // isolated from some of the nodes)
+        let last_block_info =
+            if let Some(latest) = block_infos.into_iter().max_by_key(|info| info.block_index) {
+                latest
+            } else {
+                return Err(RpcStatus::with_message(
+                    RpcStatusCode::INTERNAL,
+                    "no peers reachable".to_owned(),
+                ));
+            };
```
Comment (remoun) [inline mobilecoind/src/service.rs:1965]: Optional/Equivalent: Use `Option::ok_or_else`

```suggestion
            block_infos.into_iter().max_by_key(|info| info.block_index).ok_or_else(
                || RpcStatus::with_message(
                    RpcStatusCode::INTERNAL,
                    "no peers reachable".to_owned(),
                ))?;
```

**Review by remoun — COMMENTED** [review]
LGTM

File: connection/src/traits.rs (hunk 3 above)
Comment (garbageslam) [reply connection/src/traits.rs:127]: thanks, will apply these fixes

**Pull request merged**